### PR TITLE
[Validator] Support enums in Choice constraints

### PR DESCRIPTION
**Pull request by bpolaszek**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      |no
| New feature?  | yes
| Deprecations? | no
| Tickets       |
| License       | MIT
| Doc PR        | _TODO_

This PR allows using PHP 8.1 (backed) enums in `Choice` constraints.

Usage:

```php
enum Suit: string {
  case Hearts = 'H';
  case Diamonds = 'D';
  case Clubs = 'C';
  case Spades = 'S';
}
```

```php
class Card {
    public Suit $suit;
}
```

```php
class CardDTO {

    #[Choice(choices: Suit::class)]
    public string $suit;
}
```

### CODE REVIEW

**Comment by bpolaszek** [comment]
Regarding CI:

- Phpunit fails on PHP8.1, but that seems unrelated
- ~Psalm yells about the `BackedEnum` class - I'll require some help fixing this~

**Comment by stof** [comment]
>     * Psalm yells about the `BackedEnum` class - I'll require some help fixing this

Ignore it for now. That will be fixed in the future when running Psalm on PHP 8.1. And it won't impact the future PRs, as the CI job uses the existing codebase as the baseline.

File: src/Symfony/Component/Validator/Constraints/ChoiceValidator.php [hunk 1]
```diff
@@ -35,7 +35,7 @@ public function validate($value, Constraint $constraint)
             throw new UnexpectedTypeException($constraint, Choice::class);
         }
 
-        if (!\is_array($constraint->choices) && !$constraint->callback) {
+        if (!\is_array($constraint->choices) && !is_a($constraint->choices, \BackedEnum::class, true) && !$constraint->callback) {
```
Comment (stof) [inline src/Symfony/Component/Validator/Constraints/ChoiceValidator.php]: The `is_a` should be done only when `$constraint->choices` is a string, to avoid issues.

**Comment by derrabus** [comment]
We could silence this error in psalm.xml though.

**Comment by derrabus** [comment]
#43050

**Comment by derrabus** [comment]
If you perform a rebase on 5.4, Psalm should not complain about the enum classes anymore.

File: src/Symfony/Component/Validator/Constraints/ChoiceValidator.php (hunk 1 above)
Comment (bpolaszek) [reply src/Symfony/Component/Validator/Constraints/ChoiceValidator.php]: I wrapped the condition in a variable so that it's more readable.

**Comment by bpolaszek** [comment]
> If you perform a rebase on 5.4, Psalm should not complain about the enum classes anymore.

Thank you @derrabus - it works 🙂

**Comment by derrabus** [comment]
Can you elaborate, when would I use this validator?

Your example was:

```php
class CardDTO
{
    #[Choice(choices: Suit::class)]
    public string $suit;
}
```

But my understanding of enums is that they're type-safe. Why would I prefer a string field over using the actual enum type like in your other example:

```php
class Card {
    public Suit $suit;
}
```

Alternatively:

```php
class Card {
    #[Type(Suit::class)]
    public $suit;
}
```

**Comment by stof** [comment]
@derrabus my guess is that this would be used on a string field in a DTO receiving user input (which cannot be the enum instance and can contain invalid data, which is the whole point of using the validator), for which you want to make sure that you can use `Suit::from($cardDto->suit)` when converting the validated DTO to the domain object.

**Comment by ro0NL** [comment]
We could argue hydrating raw payloads/userdata into objects generally, i believe. The alternative route to me is #40830 + #42502, thus hydrating valid state, or validating invalid hydrated state.

From a DIY POV, this makes sense. (you need the constraint for raw payloads either way, if decoupled from Serializer).

**Comment by bpolaszek** [comment]
Yes, @stof summed it up well. If you try to unserialize an invalid value to the enum property directly, the serializer won't be happy and you'll end up with either a 500 or a 400 error.

By pre-validating it through a DTO, you keep 1 source of truth about the valid values, and you can get a more accurate constraint validation error which will tell you the exact path where it failed. Besides, if you just type-hint your property as `string` with `Choice` the old way, you still don't have 100% certainty that your property cannot have an unexpected value, what is the exact point of using enums.

Here's another example, like altering a `Post` entity with Api-Platform:
```php
enum PostStatus: int {
    case DRAFT = 0;
    case PUBLISHED = 1;
    case ARCHIVED = 2;
}

#[ApiResource(input: PostInput::class)]
class Post {
    public Ulid $id;
    public string $title;
    public PostStatus $status;
}

class PostInput {
    public string $title;
    #[Choice(choices: PostStatus::class)]
    public int $status;
}
```

The following request:
```http
PUT /posts/01FEB4F1F374ZZWVD1XWHS2SEM
Content-Type: application/json

{"title": "foo", "status": 42}
```

would get the following _HTTP 422 unprocessable entity_ response:
```json
{
  "@context": "\/api\/contexts\/ConstraintViolationList",
  "@type": "ConstraintViolationList",
  "hydra:title": "An error occurred",
  "hydra:description": "status: The value you selected is not a valid choice.",
  "violations": [
    {
      "propertyPath": "status",
      "message": "The value you selected is not a valid choice.",
      "code": "8e179f1b-97aa-4560-a02f-2a8b42e49df7"
    }
  ]
}
```

**Comment by bpolaszek** [comment]
@ro0NL  I didn't notice the existence of @lyrixx's #42502 - it looks indeed that we addressed similar issues :-) 

However, I think his approach is more like a global setting you would apply to your whole API, whereas mine targets specific properties usable in DTO contexts. 

Another use case, with the `Choice` validator allowing multiple values, you can unserialize an array of strings resulting in valid enums, i.e.

```php
enum Feature: string {
   case LIGHT_DETECTION = 'light';
   case MOVEMENT_DETECTION = 'movement';
}

#[ApiResource(input: DeviceInput::class)]
class Device {
    /**
     * @var array<Feature>
     */
    public array $features;
}

class DeviceInput {
    /**
     * @var array<string>
     */
    #[Choice(choices: Feature::class, multiple: true)]
    public array $features;
}
```

**Comment by ro0NL** [comment]
AFAIK, what we pursue is

```php
#[DeserializeFromRequest(request, #[Assert\Collection(...)])
controller(Infra\Request $request, Domain\Data $domainObject): mixed {
    try {
        return $domainObject->dataOfInterest; // typed and serializable
    } catch (ConstraintViolations $e) {
        throw $e; // typed, serializable and translatable
    }
}
```

and be done. In that sense, to me the "controller" is the source of truth.

Transforming validated payload is another interesting area, e.g. ID => entity

File: src/Symfony/Component/Validator/Test/SampleBackedEnum.php [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Test/SampleBackedEnum.php]: should be removed

File: src/Symfony/Component/Validator/Test/SampleBackedEnum.php [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+namespace Symfony\Component\Validator\Test;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Test/SampleBackedEnum.php]: this should be in the `Tests` namespace

File: src/Symfony/Component/Validator/Constraints/ChoiceValidator.php [hunk 4]
```diff
@@ -55,6 +59,13 @@ public function validate($value, Constraint $constraint)
                 throw new ConstraintDefinitionException('The Choice constraint expects a valid callback.');
             }
             $choices = $choices();
+        } elseif (is_a($constraint->choices, \BackedEnum::class, true)) {
+            $choices = array_map(
+                function (\BackedEnum $enum) {
+                    return $enum->value;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Constraints/ChoiceValidator.php]: you might be able to use array_column instead

File: src/Symfony/Component/Validator/Constraints/ChoiceValidator.php [hunk 5]
```diff
@@ -55,6 +59,13 @@ public function validate($value, Constraint $constraint)
...
+                },
+                $constraint->choices::cases()
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Constraints/ChoiceValidator.php]: this might fail when choices is a string, isn't it?
does this mean there might be a missing test case?

Comment (stof) [reply src/Symfony/Component/Validator/Constraints/ChoiceValidator.php]: when reaching this, `$constraint->choices` is a string containing a class name, which is then working to call the static method.

**Comment by derrabus** [comment]
> The alternative route to me is #40830 + #42502

That would've been my idea as well. This is a serialization/hydration issue. Storing the enum instance as a string defeats the purpose of enums.

*If* we want to pursue this approach anyway, I'd suggest to introduce a new property (e.g. `$enum`) instead of reusing the `$choices` property.

File: src/Symfony/Component/Validator/Constraints/ChoiceValidator.php (hunk 4 above)
Comment (bpolaszek) [reply src/Symfony/Component/Validator/Constraints/ChoiceValidator.php]: TIL we can use `array_column` with an array of objects! 🤯😍

**Comment by bpolaszek** [comment]
There's no such thing as an enum "instance" - my only options to refer to an enum's cases are `Enum::class` and `Enum::cases()`, the latter one being prohibited by annotations/attributes. 

Since an enum actually _is_ a choice (and something more exhaustive / reliable than an array), I don't really get the point of not being able to use it in a choice constraint.

I didn't think about a `$enum` property, but it makes sense indeed, since using a callable already requires a separate property. I just need to know the precedence to follow (callback -> enum -> choices ?) - if this matters. Should I proceed?

**Comment by derrabus** [comment]
> There's no such thing as an enum "instance"

https://3v4l.org/5jH8Q/rfc#vgit.master

**Comment by bpolaszek** [comment]
> > There's no such thing as an enum "instance"
> 
> https://3v4l.org/5jH8Q/rfc#vgit.master

There's a misundertanding - the "instance" here is one of the available cases, not a list of all allowed cases (which is what we want for the `Choice` constraint). Thought you were speaking about some sort of object instance or callable where we could iterate over all cases.

**Comment by derrabus** [comment]
> the "instance" here is one of the available cases,

Correct and each case is an instance of the enum. We can use PHP's type system to check this.

> not a list of all allowed cases (which is what we want for the `Choice` constraint).

No, what you define as cases here are actually the scalar values that are used to back the enum. My understanding of enums is that the backed value exists for storage and transport serialization. Deserializing an object with an enum should yield an instance of the enum class and not the backed value. That's what I meant when I wrote that this is a serialization/hydration concern.

* The serializer should already be able to restore an enum from its backed value.
* The form component should be patched to use enums in choice fields.
* Doctrine ORM should be patched to hydrate an enum from a `VARCHAR` field.
* …

In other words: enums exists so we don't need to perform this kind of validation anymore.

Don't get me wrong, I won't block this PR. All I'm saying is that I don't feel we need it.

**Comment by ro0NL** [comment]
@derrabus IMHO validator constraints provide i18n friendly error messages, in APIs, forms, everywhere. In that sense, i'm waiting for a `ValidatingDeserializer` (bridging both components) and have validation before hydration.

**Comment by bpolaszek** [comment]
> In other words: enums exists so we don't need to perform this kind of validation anymore.

Enums currently address a single value - but it's quite easy to validate an array of values against an enum, if you want to keep 1 source of truth for allowed values (and benefit from static analysis when processing values).
However you have to type it as an array - that's why I thought that one-liner would do the job.

```php
class DeviceInput {
    /**
     * @var array<string>
     */
    #[Choice(choices: Feature::class, multiple: true)]
    public array $features;
}
```

**Comment by derrabus** [comment]
Again: validate the enum, not the backed value.

```php
class DeviceInput {
    /**
     * @var Feature[]
     */
    #[All(new Type(Feature::class))]
    public array $features;
}

```

**Comment by derrabus** [comment]
> validator constraints provide i18n friendly error messages

I see your point here. But that means we're implementing workarounds for missing serializer features into the validator. Again, we can do that, it's just that my gut feeling tells me we're working on the inferior solution.

Can we make the error raised by the serializer in such cases easier to handle?

> I'm waiting for a `ValidatingDeserializer` (bridging both components) and have validation before hydration.

Do you have an idea how to make that happen?

**Comment by derrabus** [comment]
Related: I have implemented enum choice support for the form component: #43095.

**Comment by ro0NL** [comment]
> > validator constraints provide i18n friendly error messages
>
> I see your point here. But that means we're implementing workarounds for missing serializer features into the validator. 

Well, why duplicate the effort? Serializer could know about Validator IMHO. #42502 may add some general purpose value, but i believe it's not as nearly as powerful as Validator can be.

> > I'm waiting for a ValidatingDeserializer (bridging both components) and have validation before hydration.
>
> Do you have an idea how to make that happen?

Just a ValidatingDeserializer/"validation before hydration" conceptually, and https://github.com/symfony/symfony/pull/43047#issuecomment-921244890 for glueing :)

**Comment by bpolaszek** [comment]
Alright, I get your points. 
Can I safely close this PR?

**Comment by fabpot** [comment]
Closing this PR after reading the comments again. Thank you for proposing.

**Comment by goodjinny** [comment]
According to #44576, which was closed as a duplication of current PR. 

I fully agree with the arguments about using validator components with serializer/form components. But what about using Validator as an independent component? Let's assume we develop some console application where we don't need to use serializer and form components, but need validation of user input. So we use only console and validator components.
And in this case, there is no ability to validate some possible dto/entity objects using \BackedEnum as a storage of validation choices. So, the feature like that would be very convenient.

**Comment by bpolaszek** [comment]
> According to #44576, which was closed as a duplication of current PR.
> 
> I fully agree with the arguments about using validator components with serializer/form components. But what about using Validator as an independent component? Let's assume we develop some console application where we don't need to use serializer and form components, but need validation of user input. So we use only console and validator components. And in this case, there is no ability to validate some possible dto/entity objects using \BackedEnum as a storage of validation choices. So, the feature like that would be very convenient.

I'm using the Validator into CLI apps myself and I didn't even think about that! 🙂

**Comment by goodjinny** [comment]
> I'm using the Validator into CLI apps myself and I didn't even think about that! slightly_smiling_face

I understand. We didn't have Enum's before, so we built our apps based on other features. But since PHP 8.1 was released, enums exist.

**Comment by derrabus** [comment]
> But what about using Validator as an independent component?

Then it's still a serialization/hydration concern to me. But fine, let's explore that argument further.

> And in this case, there is no ability to validate some possible dto/entity objects using \BackedEnum as a storage of validation choices.

Okay, questions:

* How do you turn user input into a DTO in this case?
* Does that DTO of yours store an _actual_ enum case or just its string representation?

**Comment by goodjinny** [comment]
> * How do you turn user input into a DTO in this case?
> 
> * Does that DTO of yours store an _actual_ enum case or just its string representation?

Ok, agree. In this case, it's better to use an enum typed property in your dto/entity. But what about if we want to simplify our console app logic and not use DTO at all?

For example:
```php
....
        $validInput = $io->ask('Do some input', 'default', function ($input) {
            $errors = $this->validator->validate(
                $input,
                new Choice(...)
            );

            If ($errors->count()) {
                throw new \RuntimeException('Input is wrong');
            }

            return $input;
        });
...
```
If possible choices are already stored in \BackedEnum, it would be convenient to pass this enum directly
```php
            $errors = $this->validator->validate(
                $input,
                new Choice(SomeEnum::class)
            );
```

**Comment by ro0NL** [comment]
> > But what about using Validator as an independent component?

> Then it's still a serialization/hydration concern to me

The i18n friendly error message is still a validator concern to me.

**Comment by derrabus** [comment]
> For example:
> 
> ```php
> ....
>         $validInput = $io->ask('Do some input', 'default', function ($input) {
>             $errors = $this->validator->validate(
>                 $input,
>                 new Choice(...)
>             );
> 
>             If ($errors->count()) {
>                 throw new \RuntimeException('Input is wrong');
>             }
> 
>             return $input;
>         });
> ...
> ```

This is an excellent example: `$validInput` at this point is still a string. Why? **If** you created an `enum` for all valid inputs, why don't you hydrate an enum case from the input?

```php
....
        $validInput = $io->ask('Do some input', 'default', function ($input) {
            if (null === $input) {
                return null;
            }

            try {
                return YourEnumHere::from($input);
            } catch (\ValueError $e) {
                throw new \RuntimeException('Input is wrong', previous: $e);
            }
        });
...
```

**Comment by goodjinny** [comment]
> This is an excellent example: `$validInput` at this point is still a string. Why? **If** you created an `enum` for all valid inputs, why don't you hydrate an enum case from the input?

Indeed, your solution looks more elegant and logical. Don't have other arguments at the moment.

**Comment by einenlum** [comment]
I read the whole conversation and I somewhat disagree with this decision.

Some people say you shouldn't store the string representation in your DTO but the BackedEnum itself. I agree, but sometimes you don't really have the choice. Some comments said it's a serialization problem and hence, it's not a validator issue (especially since Symfony Serializer handles well deserializing enums and providing a good validation output).
The problem is, not everyone uses the Symfony Serializer. Some people use different serializers, like JMS (which until now does not handle enums). Some people want to use a DTO created from several CLI arguments.

I understand we can use a try/catch and throw an exception but it is still related to user input and therefore should IMO be treated the same way as if we use an choice array validator. A user not providing a correct value in a set of choices should not get a different response whether it's been implemented as a choice array or a backed enum in the backend.

If we want right now to get the same behavior (and validate all the user input at once) we must use a very cumbersome way:

```php
$errors = $this->validator->validate($dto);

try {
    Enum1::from($dto->enum1);
} catch (\ValueError|\TypeError $e) { // We also need to catch TypeError if the value is null
    $choiceConstraint = new Choice();
    $validChoices = array_map(fn(Enum1 $enum) => $enum->value, Enum1::cases());
    $errors->add(new ConstraintViolation(
        $choiceConstraint->message,
        $choiceConstraint->message,
        $validChoices,
        $dto,
        'enum1',
        $dto->enum1
    ));
}

try {
    Enum2::from($dto->enum2);
} catch (\ValueError|\TypeError $e) {
    $choiceConstraint = new Choice();
    $validChoices = array_map(fn(Enum2 $enum) => $enum->value, Enum2::cases());
    $errors->add(new ConstraintViolation(
        $choiceConstraint->message,
        $choiceConstraint->message,
        $validChoices,
        $dto,
        'enum2',
        $dto->enum2
    ));
}

if (count($errors)) {
    // ...
}
```

I really understand the fact that in the best possible world we get a DTO already hydrated with a backed enum, but not everyone uses the full Symfony stack. Enums are a perfect representation of a choice (substituting years of shaky practice regarding using constants and abstract classes) and not providing them when validating a choice sounds really weird to me. [Laravel has a validation rule for them](https://laravel.com/docs/8.x/validation#rule-enum). I'm not saying it's a good argument, but it means that many PHP users will expect (as I did) that it's possible on Symfony and will be troubled by this.

If Symfony maintainers still don't want to have this feature, I think it should at least be explained in the Constraints/Choice documentation, because I think there will be a lot of questions and pull requests regarding this potential feature in the next months/years.

**Comment by derrabus** [comment]
> Some people use different serializers, like JMS (which until now does not handle enums).

I understand that, but tell me: Wouldn't it be the better solution to patch JMS Serializer instead?

**Comment by einenlum** [comment]
@derrabus I agree with you on this specific point. I'm [working on this](https://github.com/schmittjoh/serializer/issues/1373#issuecomment-1002676386).

**Comment by HeahDude** [comment]
Coming late on this, but I don't really understand the arguments here.
It looks like the main reason for closing a lot of related issues/PRs is because the enum could avoid the need of the `Choice` constraint in the first place.

I don't agree that this is enough when there is still a need for using the `Choice` constraint.
If I need the `Choice` constraint on a property it's not because I want/can use a typed enum instead nor deal with the extra complexity of trying to transform it.
Yet using an enum to define a set of choices can also provide many benefits since such enum can be used elsewhere and in contexts presented by everyone here as well, and would prevent IMHO to overengineer something quite simple.

Unless we choose to deprecate the `Choice` constraint in favor of native PHP enums, I'd say we should not be against improving its DX, unless it's too much complexity in the core to be considered.

**Comment by derrabus** [comment]
> It looks like the main reason for closing a lot of related issues/PRs is because the enum could avoid the need of the `Choice` constraint in the first place.

Kind of, yes.

* If you want to validate a property that should hold an enum case, there's the `Type` constraint that does that.
* If hydrating the enum case from its backed values is too complicated, fix your hydrator (e.g. Symfony Form, Symfony Serializer and Doctrine ORM do support this now).
* If you don't want to hydrate the enum case at all, you don't need an enum and the current `Choice` constraint will work perfectly fine for you.

> I don't agree that this is enough when there is still a need for using the `Choice` constraint.

You're trying to disprove an argument nobody makes. The `Choice` is absolutely useful, nobody wants to remove it.

If you read and understood all of this and still think you _need_ this feature, please describe your use-case and we'll talk about it.

**Comment by HeahDude** [comment]
>You're trying to disprove an argument nobody makes. The Choice is absolutely useful, nobody wants to remove it.

>If you don't want to hydrate the enum case at all, you don't need an enum and the current Choice constraint will work perfectly fine for you

I you agree with this why don't you accept that everyone wants to set the choices with enums 🤔, they are basically designed for such use case.

>If you want to validate a property that should hold an enum case, there's the Type constraint that does that.

It looks like a poor alternative to me since the default message will tell that value X is not typed (enum) Y.
Also the Type constraint makes it a pain to implode cases in mapping depending on the chosen format (i.e other than native PHP).
Whereas the Choice constraint tells a value is not among a given set of values which is what we want by default.

>If you read and understood all of this and still think you need this feature, please describe your use-case and we'll talk about it.

The use a case is simple: a string needs to be validated against a set of values. There is no hydration, no serialization, no form, just a string validated among other values validated with the validator.
Changing the type is not a way to go since it introduces complexity instead of letting a constraint doing the job as it should.

**Comment by stof** [comment]
> I you agree with this why don't you accept that everyone wants to set the choices with enums thinking, they are basically designed for such use case.

Enums are designed to be hydrated. They are not designed to work with the backing value all the time.

> just a string validated among other values validated with the validator.

and that's not what PHP enums are. They are **not** strings but objects.

**Comment by HeahDude** [comment]
>Enums are designed to be hydrated. They are not designed to work with the backing value all the time.

Yes of course, the scope of this feature is the backed enum (string or int), which by design is always backed, and fits the needs to use iterable cases to define an immutable set of valid choices for string or int values to be validated. The enum can then also be safely hydrated in other layers if needed.

**Comment by ThomasLandauer** [comment]
> If you want to validate a property that should hold an enum case, there's the `Type` constraint that does that.

So the recommended way for what used to be

```php
#[Assert\Choice(choices: ['foo','bar'], message: 'Choose either foo or bar')]
```
...is now?:

```php
#[Assert\Type(type: FooBarEnum::class, message: 'Choose either foo or bar'))]
```
I wouldn't have expected that, so I'd like to add an enum example to https://symfony.com/doc/current/reference/constraints/Type.html and even a note about enums at https://symfony.com/doc/current/reference/constraints/Choice.html
Good idea?

**Comment by derrabus** [comment]
> ```php
> #[Assert\Choice(choices: ['foo','bar'], message: 'Choose either foo or bar')]
> ```

If you want to validate a string against a list of allowed strings, this is still the way to go.

**Comment by bpolaszek** [comment]
> > ```
> > #[Assert\Choice(choices: ['foo','bar'], message: 'Choose either foo or bar')]
> > ```
> 
> If you want to validate a string against a list of allowed strings, this is still the way to go.

Enum has the advantage of single source of truth, IMO

**Comment by fancyweb** [comment]
Personally, I've already used:
```php
new Choice([
    'choices' => array_map(
        static fn (MyEnum $myEnum): string => $myEnum->value,
        MyEnum::cases(),
    ),
])
```
many times already and I'd prefer to be able to use the enum class directly.

**Comment by derrabus** [comment]
@fancyweb Do you actually hydrate the enum elsewhere, e.g. by calling `MyEnum:from(…)`?

**Comment by fancyweb** [comment]
Yes, but once the value has been validated.

My personal use case is:
1. manually `json_decode` the request content
2. validate that whole array with the `Collection` constraint
3. manually instantiate an object for further usage

So before PHP 8.1:
```php
$content = json_decode($request->getContent(), true);

$this->myCustomValidatorService->validate($content, new Collection([
    'foo' => new Choice([MyFakeEnum::FOO, MyFakeEnum::BAR]),
]));

$obj = new MyClass($content['foo']); // MyClass::$foo type is string
```

What I expected on PHP 8.1:
```php
$content = json_decode($request->getContent(), true);

$this->myCustomValidatorService->validate($content, new Collection([
    'foo' => new Choice(MyRealEnum::class),
]));

$obj = new MyClass(MyRealEnum::from($content['foo'])); // MyClass::$foo type is MyRealEnum
```

The real code is obviously way more complicated than that, I summarized it.

`myCustomValidatorService` kind of "rethrows" Symfony's constraint violations as an exception. In my use case, providing an unexpected value is a bad request error, I want invalid values to be treated as an "invalid choice" just like before.

I'm not saying the existing code is state-of-the-art or anything but it works and I won't personally re-architecture it just to be able to use `MyEnum::from()` directly without any validation. I just wanted to use the `MyRealEnum` type to strengthen the code (`string` -> `MyRealEnum` in properties) and to have one source of truth for the input valid values.

Personally, I'm +0 on this PR, I understand arguments from both sides and I'm fine using the workaround. I guess it really depends on how the app is structured.

**Comment by AlexeyKosov** [comment]
What did this end up with? Will we get the ability to use `Enum`s with the `Choice` validator?

**Comment by stof** [comment]
@AlexeyKosov if you want to check that the property is one of the instances of the enum, you can already use `@Assert\Choice(callback="MyEnum::cases")` (but then, asserting the type as being `MyEnum` is simpler).
If you want to assert that the value is one of the backing value of the enum, we don't have a shortcut for that in core (you will need to implement the callback yourselves, or pass the choices in a different way)

**Comment by HeahDude** [comment]
So can we consider reopening such PR for 6.2 🙏 ? Or shall we lock this discussion as "won't fix"?

**Comment by stof** [comment]
This one has been rejected by several members of the core team. So it has no chance to pass for 6.2 IMO (we would have to *all* change our mind). Please read the discussion above for the arguments

**Comment by nicolas-grekas** [comment]
Adding this method to backed enum classes could be a nice way to go for ppl that really want to deal with backed values directly:

```php
public static function values(): array
{
    return array_column(self::cases(), 'value');
}
```

We could even suggest making this the default implementation provided by the engine. /cc @Crell

**Comment by HeahDude** [comment]
For the record https://github.com/symfony/symfony/discussions/47691#discussioncomment-3741015.

Very not convenient, this PR should really be reconsidered.

**Comment by Tobion** [comment]
I also think we should re-consider this. While I agree that in most cases it would be cleaner to handle this in the serialization directly, I don't think Symfony should enforce this by not supporting enums in the validator.
People are coming from different directions and maybe they don't want to or cannot handle this in the serialization.

To me it is unexpected that I can define my choices with an array of constants but not with an enum. Symfony and the individual components should stay flexible and this feature request would just be a few lines of code. So it improves ease of use by just adding a tiny feature (even if not everybody agrees that this is the best approach).

**Comment by derrabus** [comment]
> by not supporting enums in the validator.

I'm fine with revisiting the topic, but please let's get this straight: The validator _does_ support enums, via the `Type` or `Choice` constraints. It always has. This PR on this other hand has never been about validating enums, but about validating if a given value is a possible backing value of a specific enum.

Unlike other languages and unlike people had been building enums before PHP supported them, PHP makes a big difference between an enum and its backed value and we should do the same unless we want to cause even more confusion.

If we want to implement this kind of validation:

1. The constraint declaration should make it absolutely clear that we're not validating an enum, e.g. make it a new constraint and call it `BackedEnumValue` or something like that.
2. Such a constraint must not accept an enum case, only backed values.
3. It should come with a piece of documentation explaining the difference between `Type(MyEnum::class)` and the new constraint and why that new constraint isn't always the best option.

**Comment by vincent-joignie-dd** [comment]
If anyone else is still searching how to handle `BackedEnums` with `Choices` validator:
From few releases, Doctrine is nicely handling `BackedEnum` as a direct property type, so I'm using it like this now:

```php
enum MyEnum: string
{
    case ValueOne = 'one';
    case ValueTwo = 'two';
    case ValueThree = 'three';
}

class MyClass 
{
    #[Assert\NotBlank]
    #[Assert\Choice(callback: [MyEnum::class, 'cases'])]
    #[Column(type: 'string', /* blabla... */)]
    private MyEnum $property;
}
```

PS: I'm using this with [ApiPlatform](https://github.com/api-platform/api-platform) and everything is working perfectly.

**Comment by derrabus** [comment]
> ```php
> #[Assert\Choice(callback: [MyEnum::class, 'cases'])]
> private MyEnum $property;
> ```

That constraint is kind-of pointless because the native property type `MyEnum` would cause a `TypeError` if you tried to set that property to anything that is not part of what `MyEnum::cases()` returns.

**Comment by vincent-joignie-dd** [comment]
> That constraint is kind-of pointless because the native property type MyEnum would cause a TypeError if you tried to set that property to anything that is not part of what MyEnum::cases() returns.

You're right, but in a form validation context (or an APIPlatform write (POST, PATCH, PUT) context) it avoids to skip the validation context and throw a brutal exception. With it you stay "on the road" of validating all your data without letting "unpredictable" exception to be thrown!

In the end you're kind of right, because with a good exception management this should not crash anything. IMO it's more like a good practice to check all possible values of a field, in order to be as restrictive as possible!

**Comment by xabbuh** [comment]
> You're right, but in a form validation context (or an APIPlatform write (POST, PATCH, PUT) context) it avoids to skip the validation context and throw a brutal exception. With it you stay "on the road" of validating all your data without letting "unpredictable" exception to be thrown!

I don't understand how that would work as the value is mapped first (and thus the exception being thrown) before validation happens.

**Comment by derrabus** [comment]
Anyway, I'd like to lock this discussion. This PR usually pops up if people google for how to validate enums in Symfony and I'd like to keep it a helpful resource in that regard. New comments on this PR just add more noise and confusion.

This PR was about validating if a string value is a valid backing value for an enum. In [this comment](https://github.com/symfony/symfony/pull/43047#issuecomment-1107930895), I summed up why I felt that the validator is the wrong place for that:

> * If you want to validate a property that should hold an enum case, there's the `Type` constraint that does that.
>
> * If hydrating the enum case from its backed values is too complicated, fix your hydrator (e.g. Symfony Form, Symfony Serializer and Doctrine ORM do support this now).
>
> * If you don't want to hydrate the enum case at all, you don't need an enum and the current `Choice` constraint will work perfectly fine for you.

If you still want to validate strings this way, [Nicolas has shared a snippet](https://github.com/symfony/symfony/pull/43047#issuecomment-1161704955) for a function that you can use as a callback for the Choice constraint:

> Adding this method to backed enum classes could be a nice way to go for ppl that really want to deal with backed values directly:
> 
> ```php
> public static function values(): array
> {
>     return array_column(self::cases(), 'value');
> }
> ```

Anyone who still wants to create a dedicated constraint for this kind of validation, please read [this comment](https://github.com/symfony/symfony/pull/43047#issuecomment-1366930443) first.

**Comment by derrabus** [comment]
A new attempt to solve this issue has been made in #54226.
